docker-storage-setup.sh

```diff
@@ -111,12 +111,51 @@ get_deferred_deletion_string() {
   [ -z "$major" ] && return 0
   [ -z "$minor" ] && return 0
 
-  # docker 1.9 onwards supports deferred device deletion. Enable it.
-  if [ $major -gt 1 ] || ([ $major -eq 1 ] && [ $minor -ge 9 ]);then
-    echo "--storage-opt dm.use_deferred_deletion=true"
+  if should_enable_deferred_deletion $major $minor; then
+     echo "--storage-opt dm.use_deferred_deletion=true"
   fi
 }
 
+should_enable_deferred_deletion() {
+   # docker 1.9 onwards supports deferred device deletion. Enable it.
+   if [ $1 -lt 1 ] || ([ $1 -eq 1 ] && [ $2 -lt 9 ]);then
+      return 1
+   fi
+   if platform_supports_deferred_deletion; then
+      return 0
+   fi
+   return 1
+}
+
+platform_supports_deferred_deletion() {
+        local deferred_deletion_not_supported=false
+        local pipe1=/tmp/fifo1
+        local pipe2=/tmp/fifo2
+        mkfifo $pipe1
+        mkfifo $pipe2
+        mkdir /tmp/foo
+        mount -o bind /tmp/foo /tmp/foo
+        unshare -m ./dss-child-read-write.sh $pipe1 $pipe2 &
+        read n < $pipe1
+        umount /tmp/foo
+        if !(rmdir /tmp/foo) 2>/dev/null; then
+            deferred_deletion_not_supported=true
+        fi
+        echo "pipe2" > $pipe2
+        trap cleanup EXIT
+        if [ "$deferred_deletion_not_supported" = true ];then
+            return 1
+        else
+	    return 0
+        fi
+}
+
```

Comment:
Can we simplify structure a bit. May be something like as follows.

local deferred_removal_supported

rmdir /tmp/foo > /dev/null
deferred_removal_supported = $?

<Do the cleanup>

return $deferred_removal_supported